ddtrace/ddtrace.go

```diff
@@ -43,6 +43,7 @@ type Tracer interface {
 type Span interface {
 	opentracing.Span
 	FinishWithOptionsExt(opts ...FinishOption)
+	ContextExt() SpanContextExt
```

Comment:
I do not like this design. I would rather try extending it more "dynamically" like:

```
func SpanID(s Span) uint64 {
   ext, ok := s.(interface{ SpanID() uint64 })
   if !ok {
       return 0
   }
   return ext.SpanID()
}
```

something like: https://github.com/golang/go/blob/master/src/errors/wrap.go

it would not force us to do changes like e.g. in `NoopSpanContext`

Reply:
What happens when this function returns zero? Instrumentation will have to skip the entire logic that depends on valid span ID being returned or use 0 as the key both of which make the instrumentation code not so ideal. Extending the inteface allows us a guaranteed result from `ContextExt().SpanId()`. Plus it is an established pattern with `FinishWithOptionsExt()` in the repo.

Reply:
Will take a closer look later. Do you need it today or can I deep dive tomorrow?

Reply:
No one else is supposed to implement this interface so there isn't really any downside to this. It might _feel_ ugly but there really isn't any added complexity or unreliability here. Ideally I'd have liked to extend the `SpanContext` interface to add methods to that but that would make our Span interface incompatible with OpenTracing. 

Another thing I'd have liked was to add `SpanID()` and `TraceID()` method to the span interface directly. This works but it doesn't help instrumentations trying to extract span ID from an extracted (remote) context. However, for that it is more acceptable to have conditional logic. I'll see we can do that instead.

Reply:
Replaced with a function call that may potentially return `0` if our implementation of ddtrace.Span is not being used. Interface extension kind of guaranteed any other implementations would take care of this. However, we don't really have to support people extending our interface and implementing custom tracer/span types and I certainly don't know any cases where people have done this. So I think we're good to go with the helper function instead.